lightning/src/onion_message/packet.rs

```diff
@@ -156,30 +158,45 @@ impl Writeable for (Payload, [u8; 32]) {
 			},
 			Payload::Receive {
 				control_tlvs: ReceiveControlTlvs::Unblinded(control_tlvs), reply_path,
+				message: Tlv { tag, value, },
 			} => {
 				let write_adapter = ChaChaPolyWriteAdapter::new(self.1, &control_tlvs);
 				encode_varint_length_prefixed_tlv!(w, {
 					(2, reply_path, option),
-					(4, write_adapter, required)
+					(4, write_adapter, required),
+					(*tag, value, vec_type)
 				})
 			},
 		}
 		Ok(())
 	}
 }
 
-// Uses the provided secret to simultaneously decode and decrypt the control TLVs.
+// Uses the provided secret to simultaneously decode and decrypt the control TLVs and data TLV.
 impl ReadableArgs<SharedSecret> for Payload {
 	fn read<R: Read>(r: &mut R, encrypted_tlvs_ss: SharedSecret) -> Result<Self, DecodeError> {
 		let v: BigSize = Readable::read(r)?;
 		let mut rd = FixedLengthReader::new(r, v.0);
 		let mut reply_path: Option<BlindedRoute> = None;
 		let mut read_adapter: Option<ChaChaPolyReadAdapter<ControlTlvs>> = None;
 		let rho = onion_utils::gen_rho_from_shared_secret(&encrypted_tlvs_ss.secret_bytes());
+		let (mut message_type, mut message_bytes) = (None, None);
+		let mut decode_custom_tlv = |msg_type, mut msg_reader: &mut FixedLengthReader<&mut &mut FixedLengthReader<&mut R>>| {
+			if msg_type >= 64 {
+				// Don't allow reading more than one data TLV from an onion message.
+				if message_bytes.is_some() || message_type.is_some() {
+					return Err(DecodeError::InvalidValue)
+				}
+				message_type = Some(msg_type);
+				decode_tlv!(msg_reader, message_bytes, vec_type);
+				return Ok(true)
```

Comment:
@jkczyz Should this bool be an enum? Should we get rid of the result and always return a trinary enum? Let me know what you prefer

Reply:
I had been thinking the custom decoder could signal an unknown type by returning a variant of `DecodeError` (possibly a new variant). This would presumed to be returned only if none of the bytes were consumed (i.e., only when checking the type). Otherwise, another custom decoder could not process it if the previous one consumed some bytes. If we make a trait for the custom decoder, we could have a separate method for checking if it understands the type and another method for decoding, perhaps?

Reply:
Ah, I think that I misread the `?` repetition as `*` and thought this would handle known payloads (e.g., `invoice_request`) as a separate custom decoder. I guess logic for reading such known types would be added to the closure?

Related: Will handling be separate from decoding? If so, will we need the type still to figure out if handler for a known type is needed or a custom type, IIUC.